File: project/frontend-services/libs/device-management/mapper/src/lib/mappers/RecordMapper.ts
import {DeviceRecordDTO, RecordTypeDTO, SensorDataRecordLabelDTO,} from '@frontend-services/device-management/dto';
import {
  Device, DeviceCommand,
  DeviceInformation,
  RecordEntry,
  RecordEntryType,
  SensorDataRecordLabel,
  SubDevice
} from '@frontend-services/device-management/model';

export class RecordMapper {
  static dtoToModel(dto: DeviceRecordDTO): DeviceInformation {
    const entries = dto.entries.map((e) => {
      if (e.type == RecordTypeDTO.BASIC) {
        return new RecordEntry(e.label, e.content, RecordEntryType.BASIC);
      } else {
        if (
          (e.label as SensorDataRecordLabelDTO) ===
          SensorDataRecordLabelDTO.GPS_LONGITUDE
        ) {
          return new RecordEntry(
            SensorDataRecordLabel.GPS_LONGITUDE,
            e.content,
            RecordEntryType.SENSOR_DATA
          );
        } else {
          return new RecordEntry(
            SensorDataRecordLabel.GPS_LATITUDE,
            e.content,
            RecordEntryType.SENSOR_DATA
          );
        }
      }
    });
    const subDevices = dto.subDevices.map(deviceRef => new SubDevice(deviceRef.ref, deviceRef.id));

    const commands = dto.commands.map(com => new DeviceCommand(com.id, com.name, com.payload, com.ref, com.port));

    const device = new Device(dto.device.id, dto.device.name, dto.device.downlink);
    return new DeviceInformation(device, entries, subDevices, commands);
  }

  static modelToDto(model: DeviceInformation): DeviceRecordDTO {
    const entries = model.entries.map((e) => {
      if (e.type == RecordEntryType.BASIC) {
        return {
          label: e.label,
          content: e.content,
          type: RecordTypeDTO.BASIC,
        };
      } else {
        if (
          (e.label as SensorDataRecordLabel) ===
          SensorDataRecordLabel.GPS_LONGITUDE
        ) {
          return {
            label: SensorDataRecordLabelDTO.GPS_LONGITUDE,
            content: e.content,
            type: RecordTypeDTO.SENSOR_DATA,
          };
        } else {
          return {
            label: SensorDataRecordLabelDTO.GPS_LATITUDE,
            content: e.content,
            type: RecordTypeDTO.SENSOR_DATA,
          };
        }
      }
    });

    const subDevices = model.subDevices.map(sub => {
      return {
        id: sub.id,
        ref: sub.reference
      }
    });

    const commands = model.commands.map(com => {
      return {
        id: com.id,
        name: com.name,
        payload: com.payload,
        ref: com.ref,
        port: com.port
      }
    });

    return {
      device: {id: model.device.id, name: model.device.name, downlink: model.device.downlink},
      entries: entries,
      subDevices: subDevices,
      commands: commands
    };
  }
}

File: project/backend-services/device-management-master-backend/infrastructure/persistence/postgres/src/main/java/pt/sensae/services/device/management/master/backend/infrastructure/persistence/postgres/mapper/RecordMapper.java
package pt.sensae.services.device.management.master.backend.infrastructure.persistence.postgres.mapper;

import pt.sensae.services.device.management.master.backend.domain.model.commands.*;
import pt.sensae.services.device.management.master.backend.domain.model.device.Device;
import pt.sensae.services.device.management.master.backend.domain.model.device.DeviceDownlink;
import pt.sensae.services.device.management.master.backend.domain.model.device.DeviceId;
import pt.sensae.services.device.management.master.backend.domain.model.device.DeviceName;
import pt.sensae.services.device.management.master.backend.domain.model.records.*;
import pt.sensae.services.device.management.master.backend.domain.model.subDevices.DeviceRef;
import pt.sensae.services.device.management.master.backend.domain.model.subDevices.SubDevice;
import pt.sensae.services.device.management.master.backend.domain.model.subDevices.SubDevices;
import pt.sensae.services.device.management.master.backend.infrastructure.persistence.postgres.model.*;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

public class RecordMapper {
    public static DeviceInformationPostgres domainToPostgres(DeviceInformation records) {
        var postgres = new DeviceInformationPostgres();
        postgres.deviceId = records.device().id().value().toString();
        postgres.name = records.device().name().value();
        postgres.downlink = records.device().downlink().value();
        postgres.entries = records.records().entries().stream().map(e -> {
            var entry = new DeviceRecordEntryPostgres();
            entry.type = e instanceof BasicRecordEntry ?
                    DeviceRecordEntryTypePostgres.basic() :
                    DeviceRecordEntryTypePostgres.sensorData();
            entry.content = e.getContent();
            entry.label = e.getLabel();
            entry.records = postgres;
            return entry;
        }).collect(Collectors.toSet());

        postgres.subSensors = records.subDevices().entries().stream().map(sub -> {
            var entry = new DeviceSubSensorPostgres();
            entry.subDeviceRef = sub.ref().value();
            entry.subDeviceId = sub.id().value().toString();
            entry.controller = postgres;
            return entry;
        }).collect(Collectors.toSet());

        postgres.commands = records.commands().entries().stream().map(com -> {
            var entry = new DeviceCommandEntryPostgres();
            entry.device = postgres;
            entry.id = com.id().value();
            entry.name = com.name().value();
            entry.port = com.port().value();
            entry.payload = com.payload().value();
            entry.subDeviceRef = com.ref().value();
            return entry;
        }).collect(Collectors.toSet());

        return postgres;
    }

    public static DeviceInformation postgresToDomain(DeviceInformationPostgres records) {
        List<RecordEntry> collect = records.entries.stream()
                .map(e -> e.type.equals(DeviceRecordEntryTypePostgres.sensorData()) ?
                        new SensorDataRecordEntry(SensorDataRecordLabel.give(e.label), e.content) :
                        new BasicRecordEntry(e.label, e.content))
                .collect(Collectors.toList());

        var deviceId = new DeviceId(UUID.fromString(records.deviceId));
        var deviceName = new DeviceName(records.name);
        var deviceDownlink = new DeviceDownlink(records.downlink);

        var subDevices = records.subSensors.stream()
                .map(sub -> new SubDevice(new DeviceId(UUID.fromString(sub.subDeviceId)), new DeviceRef(sub.subDeviceRef)))
                .collect(Collectors.toSet());

        var commands = records.commands.stream()
                .map(c -> new CommandEntry(CommandId.of(c.id), CommandName.of(c.name), CommandPayload.of(c.payload), CommandPort.of(c.port), DeviceRef.of(c.subDeviceRef)))
                .collect(Collectors.toSet());

        return new DeviceInformation(new Device(deviceId, deviceName, deviceDownlink), new Records(collect), new SubDevices(subDevices), new DeviceCommands(commands));
    }
}
